mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -15,60 +16,106 @@ use std::collections::HashMap;
 use super::EthCoin;
 use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
             BalanceError, MmCoin};
+struct BalanceData {
+    ticker: String,
+    address: String,
+    balance: BigDecimal,
+}
+
+struct BalanceFetchError {
+    ticker: String,
+    address: String,
+    error: MmError<BalanceError>,
+}
+
+type BalanceResult = Result<BalanceData, BalanceFetchError>;
 
 /// This implementation differs from others, as they immediately return
 /// an error if any of the requests fails. This one completes all futures
 /// and returns their results individually.
-async fn get_all_balance_results_concurrently(
-    coin: &EthCoin,
-) -> Vec<Result<(String, BigDecimal), (String, MmError<BalanceError>)>> {
+async fn get_all_balance_results_concurrently(coin: &EthCoin) -> Vec<BalanceResult> {
     let mut tokens = coin.get_erc_tokens_infos();
-
     // Workaround for performance purposes.
     //
     // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
     // type and mapping the platform coin and the entire token list (which can grow at any time), we map
     // the platform coin to Erc20TokenInfo so that we can use the token list right away without
     // additional mapping.
     tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
-        token_address: coin.my_address,
+        token_address: Address::default(),
```

Comment:
With this approach HD account balance events will be unrecognizable from the clients. They need to know which address and the ticker balance was changed for.

Reply:
This is the token contract address, if the coin is a platform coin, it doesn't have a token contract, if it's a token but `token_ticker == coin.ticker`, `address_balance` is used for balance
https://github.com/KomodoPlatform/komodo-defi-framework/blob/86603926a2972fb94195f8693dd144e10c5bed13/mm2src/coins/eth/eth_balance_events.rs#L83-L85

Reply:
So there are 2 issues:

1- fetch_balance may use `info.token_address` to fetch balance but it returns `address` at the end.
2- When you pass `Address::default()`, this value will be used for sending balance changes in the balance events.

Reply:
I guess I don't understand the problem, maybe I am missing something. Here is my explanation of what is going on, please correct me if I am wrong

> 1- fetch_balance may use info.token_address to fetch balance but it returns address at the end.

`get_token_balance` takes the token contract address which is `info.token_address`, what is returned at the end is the user address, which is one of the addresses of the HD wallet passed to `get_all_balance_results_concurrently`.  `get_token_balance` is never used for this ticker https://github.com/KomodoPlatform/komodo-defi-framework/blob/846c22a70ab2f62d368908f0e383681679cd1736/mm2src/coins/eth/eth_balance_events.rs#L44-L47 previously (before this PR) you passed the user's address as the token address which was not right too but it was fine because it never gets used with `get_token_balance`.

> When you pass Address::default(), this value will be used for sending balance changes in the balance events. Here

`Address::default()` which is a dummy for platform coin contract address is never returned to the function sending the balance changes, the user address/s is what is returned.

Reply:
> previously (before this PR) you passed the user's address as the token address which was not right too but it was fine because it never gets used with get_token_balance.

P.S. I can't pass the user's address as token contract address as before because we now have multiple addresses per user.

Reply:
I got confused with the `address` argument of `fetch_balance` and `address` field of `Erc20TokenInfo`.. I implemented this sometime ago now it looks very confusing at the first look. Perhaps adding explanation comment on top of `Erc20TokenInfo` address would be good for the team?

Reply:
I agree that adding more comments will help, I too made a mistake before ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1539248928 because it wasn't very clear to me.

Reply:
Added some comments that explain the situation here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/dc6ef2ae1f9d9acffc86e6c73943389bd27378c2